### PR TITLE
WIP: Implement recursive shadowcast vision algorithm

### PR DESCRIPTION
**Pull request by Fuco1**
Implementation of http://www.roguebasin.com/index.php?title=FOV_using_recursive_shadowcasting vision algorithm.

This will allow us to only render visible monsters/tiles, it also supports lighting radius so we can define visible "spheres" around the player.

### CODE REVIEW

**Comment by Fuco1** [comment]
So far so good

```
[18:18:15]matus@herakleitos:~/dev/c++/freeablo-project/build
> ./shadow
...sss....
....##....
..........
.....@....
..........
..........
..######..
.ssssssss.
ssssssssss
ssssssssss
```

Now I'll have to plug it in the `GameLevel`.

**Comment by wheybags** [comment]
Nice, let me know when it's ready to merge. You might also put in that link as a comment beside the scan function?

File: apps/freeablo/faworld/gamelevel.cpp [hunk 1]
```diff
@@ -125,15 +131,17 @@ namespace FAWorld
...
 
-    void GameLevel::fillRenderState(FARender::RenderState* state)
+    void GameLevel::fillRenderState(FARender::RenderState* state, const Player& player)
     {
         state->mObjects.clear();
 
         for(size_t i = 0; i < mActors.size(); i++)
         {
-            size_t frame = mActors[i]->mFrame + mActors[i]->mPos.getDirection() * mActors[i]->getCurrentAnim()->getAnimLength();
+            if (player.canSee(mActors[i]->mPos.current())) {
```
Comment (Fuco1) [inline apps/freeablo/faworld/gamelevel.cpp]: Add a version that takes the actor as argument so we don't needlessly expose internals to another class (GameLevel)

File: apps/freeablo/faworld/player.cpp [hunk 2]
```diff
@@ -145,6 +145,11 @@ namespace FAWorld
         return Actor::getCurrentAnim();
     }
 
+    void Player::update(bool noclip, size_t ticksPassed)
```
Comment (Fuco1) [inline apps/freeablo/faworld/player.cpp]: This can be removed

File: components/level/level.cpp [hunk 3]
```diff
@@ -26,6 +26,15 @@ namespace Level
 
     const MinPillar get(size_t x, size_t y, const Level& level)
     {
+        // static std::map<std::pair<size_t, size_t>, MinPillar&> minCache;
```
Comment (Fuco1) [inline components/level/level.cpp]: Delete this commented code

File: components/vision/scanner.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
+#include "scanner.h"
+#include <iostream>
+#include <cmath>
+
+namespace Vision
+{
+
+    std::set<std::pair<int32_t, int32_t>> Scanner::getVisibleTiles()
+    {
+        mVisibleTiles = {};
```
Comment (Fuco1) [inline components/vision/scanner.cpp]: Pass this as attribute to scan

**Comment by Fuco1** [comment]
@wheybags Do you want this in 0.4?  If so maybe we can set up some milestone.  I'm scheduling a time slot over the weekend to work on this and hopefully get it into a mergable state.

Depending on how the new renderer works I might also try to implement hiding the cells that are invisible (in some rough manner at least, maybe with a switch to toggle it).  Later we can write fancy shaders for gradual darkening of the surroundings.

**Comment by wheybags** [comment]
It would be nice, but basically if it's done by then, it'll go in,
otherwise it won't.
There is already a github issue milestone for 0.4:
https://github.com/wheybags/freeablo/milestone/5
I also have a personal trello board for what I'm planning to do for 0.4
over the next few weeks: https://trello.com/b/wC2cgfrt/freeablo-summer-2017
Also, nice, looking forward to seeing it :)

**Comment by wheybags** [comment]
going to close this for now, as it's got so far out of date, but feel free to re-open later

**Pull request closed without merging**